### PR TITLE
Fix type annotation of assert functions

### PR DESCRIPTION
**Pull request by bernt-matthias**
asserts: n, delta, min, max need to be str since the allow extensions: (K|M|G|T|P|E)i

solved by automatically converting bools, integers, and suffixed integers before calling the assertion functions. 

follow up on https://github.com/galaxyproject/galaxy/pull/12528

I guess this should wait for the refactoring of the tests that is currently ongoing in https://github.com/galaxyproject/galaxy/pull/13186

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

File: lib/galaxy/tool_util/verify/asserts/archive.py [hunk 1]
```diff
@@ -49,7 +49,7 @@ def _list_from_zip(bytes, path):
     return sorted(lst)
 
 
-def assert_has_archive_member(output_bytes, path, verify_assertions_function, children, all="false", n: int = None, delta: int = 0, min: int = None, max: int = None, negate: bool = False):
+def assert_has_archive_member(output_bytes, path, verify_assertions_function, children, all="false", n: str = None, delta: str = 0, min: str = None, max: str = None, negate: bool = False):
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/verify/asserts/archive.py]: If they can be int as well this should be a Union, and if the default is None I think you need typing.Optional
```suggestion
def assert_has_archive_member(output_bytes, path, verify_assertions_function, children, all="false", n: Optional[Union[str, int]] = None, delta: Union[str, int] = 0, min: Optional[Union[str, int]] = None, max: Optional[Union[str, int]] = None, negate: bool = False):
```

File: lib/galaxy/tool_util/verify/asserts/__init__.py [hunk 2]
```diff
@@ -44,10 +51,20 @@ def verify_assertion(data, assertion_description):
...
     args = {}
     for attribute, value in assertion_description["attributes"].items():
         if attribute in assert_function_args:
-            args[attribute] = value
+            # in python 3.10 issubclass(xyz, assert_function_annotations[attribute]) should work
+            types = assert_function_annotations[attribute]
+            if get_origin(assert_function_annotations[attribute]) is not None:
+                types = get_args(assert_function_annotations[attribute])
+            if issubclass(int, types):
+                args[attribute] = parse_bytesize(value)
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/verify/asserts/__init__.py:63]: If someone specified `123M` for an integer value they'll get `123000000` as the value, or am I misunderstanding this ? Maybe you could narrow this down by creating a ByteCount type or similar ?

File: lib/galaxy/tool_util/verify/asserts/archive.py [hunk 3]
```diff
@@ -49,11 +50,19 @@ def _list_from_zip(bytes, path):
     return sorted(lst)
 
 
-def assert_has_archive_member(output_bytes, path, verify_assertions_function, children, all="false", n: int = None, delta: int = 0, min: int = None, max: int = None, negate: bool = False):
+def assert_has_archive_member(output_bytes: bytes,
+                              path: str,
+                              verify_assertions_function: Callable,
+                              children: List[etree.Element],
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/verify/asserts/archive.py]: You sure that's the only type that can be passed here ? I think this should also work with yaml test definitions

File: lib/galaxy/tool_util/verify/asserts/xml.py [hunk 4]
```diff
@@ -19,53 +20,89 @@ def assert_is_valid_xml(output):
...
     """ Asserts the specified attribute of the first element matching
     the specified path matches exactly the specified text."""
     assert_attribute_matches(output, path, attribute, re.escape(text) + "$", negate=negate)
 
 
-def assert_element_text(output, path, verify_assertions_function, children, negate: bool = False):
+def assert_element_text(output: str,
+                        path: str,
+                        verify_assertions_function: Callable,
+                        children: List[etree.Element],
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/verify/asserts/xml.py]: Again I'm not sure this isn't too tight.

File: lib/galaxy/tool_util/verify/asserts/xml.py [hunk 5]
```diff
@@ -19,53 +20,89 @@ def assert_is_valid_xml(output):
...
     """ Recursively checks the specified assertions against the text of
     the first element matching the specified path."""
     assert_xml_element(output, path, verify_assertions_function, children, negate=negate)
 
 
-def assert_xml_element(output, path, verify_assertions_function=None, children=None, attribute=None, all=False, n: int = None, delta: int = 0, min: int = None, max: int = None, negate: bool = False):
+def assert_xml_element(output: str,
+                       path: str,
+                       verify_assertions_function: Optional[Callable] = None,
+                       children: Optional[List[etree.Element]] = None,
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/verify/asserts/xml.py]: Same here

File: lib/galaxy/tool_util/verify/asserts/archive.py (hunk 3 above)
Comment (bernt-matthias) [reply lib/galaxy/tool_util/verify/asserts/archive.py]: Good point. Seems that the parser makes it a List[Dict] anyway:

https://github.com/galaxyproject/galaxy/blob/32a41700f3026aa24b151a3fd7e7205ee3fd5fee/lib/galaxy/tool_util/parser/xml.py#L741
https://github.com/galaxyproject/galaxy/blob/32a41700f3026aa24b151a3fd7e7205ee3fd5fee/lib/galaxy/tool_util/parser/yaml.py#L282

Will change it.

File: lib/galaxy/tool_util/verify/asserts/__init__.py (hunk 2 above)
Comment (bernt-matthias) [reply lib/galaxy/tool_util/verify/asserts/__init__.py:63]: Do you mean that we should not allow `123Mi` which would be `132070244352 = 123*1024^3`

**Pull request closed without merging**